### DISCUSSION TITLE
What's the return type of `bind(..).serve(...)`?

### DISCUSSION THREAD

**Discussion by luos** [tokio-rs/axum#541] [Q&A] [answered]
Hi,

I am new to Rust, sorry for this beginner question. I am trying to find out what is the return type of the `serve` method so I can put it in a function. I tried some which look like they could work but they are from private modules. 

Here is what I am trying to achieve: 

I am trying to find the return type of the `start_web_handler` function, highlighted with `??`. 

```
async fn  start_web_handler() -> Server<??AddrIncoming??, ??IntoMakeService<Router>??> {
    let app = Router::new().route("/", get(|| async { "Hello, World!" }));

    // run it with hyper on localhost:3000
    let addr = SocketAddr::from(([127, 0, 0, 1], 3000));
    let s = axum::Server::bind(&addr)
        .serve(app.into_make_service());
    s
}
```

Thanks,

**Comment by davidpdrsn** [comment, marked as answer]
From looking at the docs we see that [`serve`](https://docs.rs/hyper/0.14.15/hyper/server/struct.Builder.html#method.serve) returns a [`Server`](https://docs.rs/hyper/0.14.15/hyper/server/struct.Server.html) which implements [`Future`](https://docs.rs/hyper/0.14.15/hyper/server/struct.Server.html#impl-Future) where the `Output` is [`hyper::Result<()>`](https://docs.rs/hyper/0.14.15/hyper/type.Result.html). So if you add a `.await` inside your function the return type becomes `Result<(), hyper::Error>`.

**Reply by luos** [reply]
I guess I got a little bit confused about putting `await` into this method, as I wanted to "propagate" the future out. 
I got it working how I wanted now. Thank you!

```
pub async fn  start_listener() -> Result<(), hyper::Error> {
    let app = Router::new().route("/", get(|| async { "Hello, World!" }));

    // run it with hyper on localhost:3000
    let addr = SocketAddr::from(([127, 0, 0, 1], 3000));
    let s = axum::Server::bind(&addr)
        .serve(app.into_make_service());
    s.await
}
```

**Reply by davidpdrsn** [reply]
Right yeah that can be a little confusing. But it works since `start_listener` is itself an `async fn`. So nothing happens unless the caller `await`s the future returned by `start_listener`.

**Comment by Daniel-Xu** [comment]
Is it idiomatic to return the Server in this case? ( i need to run it later in `run_until_stop`)
![2023-03-14 at 12 55 AM](https://user-images.githubusercontent.com/548144/224771590-d56b20b9-f8f3-4eb7-a1aa-36335d0c11ab.png)

**Reply by davidpdrsn** [reply]
Sure. Why not? I mean the type is a bit big but otherwise I don't see any issues if it works.
